src/vsr/client.zig

```diff
@@ -196,52 +296,131 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
             if (self.request_timeout.fired()) self.on_request_timeout();
         }
 
-        pub fn request(
+        pub const BatchError = error{
+            // There are not enough internal resources (Messages or Demux) to allocate
+            // a batched Request.
+            TooManyOutstanding,
+        };
+
+        /// Allocate space in a vsr Message to write [event_count]StateMachine.Event(operation).
+        /// Unlike `raw_request`, a Batch will attempt to piggy-back onto an existing Message.
+        /// The returned Batch provides a slice of memory to write the events to. Once populated,
+        /// use `batch_submit` to commit the writes to the Message and send it through the Client.
+        /// NOTE: A Batch cannot be stored and must be passed to batch_submit() before calling
+        /// any other methods on the Client.
+        pub fn batch_get(
             self: *Self,
-            user_data: u128,
-            callback: Request.Callback,
             operation: StateMachine.Operation,
-            message: *Message,
-            message_body_size: usize,
-        ) void {
+            event_count: usize,
+        ) BatchError!Batch {
             assert(@intFromEnum(operation) >= constants.vsr_operations_reserved);
+            const event_size: usize = switch (operation) {
+                inline else => |operation_comptime| @sizeOf(StateMachine.Event(operation_comptime)),
+            };
 
-            self.register();
-            assert(self.request_number > 0);
+            const body_size = event_size * event_count;
+            assert(body_size <= constants.message_body_size_max);
 
-            const message_request = message.build(.request);
+            // A demux node is always reserved, either for the initial Request or a tag-along demux.
+            const demux = self.demux_pool.acquire() orelse return error.TooManyOutstanding;
+            errdefer self.demux_pool.release(demux);
+
+            demux.* = .{
+                .user_data = undefined, // Set during batch_submit().
+                .callback = undefined, // Set during batch_submit().
+                .event_count = @intCast(event_count),
+                // Assume the first demux here. Updated if enqueued onto an existing Request.
+                .event_offset = 0,
+            };
+
+            // Check-in with the StateMachine to see if this operation can even be batched.
+            // If so, find an existing Request with the same Op that has room in its Message.
+            // The request must not be the one currently inflight in VSR as its Message is sealed.
+            if (StateMachine.batch_logical_allowed.get(operation)) {
+                var it = self.request_queue.iterator_mutable();
+                while (it.next_ptr()) |request| {
+                    if (request == self.request_inflight) continue;
```

Comment:
Re: https://github.com/tigerbeetle/tigerbeetle/pull/1257#discussion_r1384075559:

> Need to avoid batching into the head Request of the queue if it's currently inflight (message has already been sent to MessageBus). The implicit contract is indeed kind of a hack.

Could we accomplish this (minus the hack) by just calling `_ = it.next_ptr();` immediately before this loop starts? (With a comment to explain.)